Amputate Links

kj4ezj opened the pull request:
From [issue 1](https://github.com/zhanghai/Untracker/issues/1), this pull request adds two simple rules that convert three different types of Google AMP links to their canonical equivalent. There may be AMP links that take other forms not covered by these rules.

> [!NOTE]
> > Google AMP (Accelerated Mobile Pages) is harmful to the free and open web for [a variety of reasons](https://old.reddit.com/r/AmputatorBot/comments/ehrq3z/why_did_i_build_amputatorbot/) that are beyond the scope of this discussion. Relevant to this app, it enables both the publisher _and_ Google to spy on readers instead of just the publisher. It also misrepresents who is serving you content in the URL bar.

There is no straightforward way I see to create a single rule covering all three of these link types using the interface exposed by Untracker, at least without running this rule on _all_ links instead of just AMP links.

## Amputate - Host
This rule converts Google AMP links of the subdomain form...
```
https://amp.example.com/some/path/news-article.html
```
...to their canonical form.
```
https://example.com/some/path/news-article.html
```
Here is my rule code.
```js
if ($.matches(url, '(.+[.])?amp[.].+')) {
    const host = $.getHost(url).replace(/(?<!\w)amp[.]/, '');
    return $.setHost(url, host);
}
```

## Amputate - Path
This rule amputates links using the path form...
```
https://example.com/amp/some/path/news-article.html
```
...or the file extension form of Google AMP...
```
https://example.com/some/path/news-article.html.amp
```
...to their canonical form.
```
https://example.com/some/path/news-article.html
```
Here is my rule code.
```js
if ($.matches(url, '.*', '.*amp.*')) {
    const path = $.getEncodedPath(url).replace(/[./]amp(?!\w)/, '');
    return $.setEncodedPath(url, path);
}
```

### Tests
I tested my changes against Untracker `v1.0.3 (4)` from [F-Droid](https://f-droid.org/packages/me.zhanghai.android.untracker).

**1. -** AMP subdomain is removed.
```
https://amp.example.com/some/path/news-article.html
```
![1](https://github.com/zhanghai/Untracker/assets/34947245/18d54aaa-64d8-425e-b363-c56e5ed9ea2a)

**2. -** Normal subdomains are preserved.
```
https://amp.news.example.com/some/path/news-article.html
```
![2](https://github.com/zhanghai/Untracker/assets/34947245/274aa1d4-27b8-4bf6-afc7-bdebbeb7a583)

**3. -** Non-AMP subdomains are not matched.
```
https://damp.example.com/some/path/news-article.html
```
![3](https://github.com/zhanghai/Untracker/assets/34947245/be89b90a-f6da-445a-baf7-d8957de3bac1)

**4. -** Non-AMP subdomains are not matched.
```
https://amperes.example.com/some/path/news-article.html
```
![4](https://github.com/zhanghai/Untracker/assets/34947245/545c11f0-84ec-4173-b7cb-dac92d3aca7a)

**5. -** AMP paths are removed.
```
https://example.com/amp/some/path/news-article.html
```
![5](https://github.com/zhanghai/Untracker/assets/34947245/8b6bebdb-0113-4be4-9474-af03c9f4eee9)

**6. -** Non-AMP paths are not matched.
```
https://example.com/damp/some/path/news-article.html
```
![6](https://github.com/zhanghai/Untracker/assets/34947245/5eb2d46b-da03-4b9f-8c04-bdd63c137ce3)

**7. -** Non-AMP paths are not matched.
```
https://example.com/ampere/some/path/news-article.html
```
![7](https://github.com/zhanghai/Untracker/assets/34947245/4efa2ee8-d6ce-4788-a886-870f6a31ea81)

**8. -** AMP file extensions are removed.
```
https://example.com/some/path/news-article.html.amp
```
![8](https://github.com/zhanghai/Untracker/assets/34947245/ca69d8a8-c553-4de8-a0ea-ffca236360de)

**9. -** AMP file extensions are removed.
```
https://example.com/some/path/news-article.amp.html
```
![9](https://github.com/zhanghai/Untracker/assets/34947245/e535a964-1a93-45ca-8403-68b92ec61dcf)

---
- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

kj4ezj commented:
Individual CLA signed.

zhanghai commented:
Is there an official way to get the original URL of an AMP page? I don't like AMP either, but having `amp` in a domain or path doesn't necessarily mean it's an AMP page since it has [a lot of other meanings](https://www.wikiwand.com/en/AMP), and I don't want to include something that may not work as a built-in rule. So I would recommend keeping it as a custom rule for now.

kj4ezj commented:
> Is there an official way to get the original URL of an AMP page?

Yes, the [AmputatorBot](https://github.com/KilledMufasa/AmputatorBot) project provides a free, public [API](https://www.postman.com/amputatorbot/workspace/amputatorbot/api/b3d8f519-6e8c-4316-9dcd-3d9d1c5b7699/collection/12422626-982a3e50-e959-442b-ab97-448b3381faa1) that uses fourteen patterns, heuristics, and can follow redirects that claims a >97% conversion rate.

Using their API instead of local rules has some implications. It would reveal the link being shared to a third-party (them), requires an Internet connection, and your app will have to wait on the API response. You may also have to mark your app on F-Droid as "depending on a non-free service" because of this.
> Everything, except for some parts of this API and the website, has been open sourced on GitHub.

I did not go this route initially because I wasn't even sure if your rule script allows me to call out to the Internet, it is more work, and these rules are sufficient in 99.9% of use-cases. I am willing to try implementing the AmputatorBot API in your app and opening another pull request, but I'm not going to spend time on it unless you are willing to accept those implications and that work would be likely to be merged.

Another option would be set the rules in this pull request to "disabled" by default. I went ahead and pushed a commit to do that so, if you want to go with this option, you can just approve or merge this pull request.

A third option still would be to dig through their codebase and implement the fourteen rules they use locally. This side-steps the implications of using their API while almost certainly addressing your concerns. However, I don't know how complicated their rules are. Again, I would be willing to work on implementing this but I won't put time into it unless you signal interest in that work.

What do you think?

zhanghai commented:
The app does allow rules to connect to Internet which is how the built-in rule for expanding short links are implemented - it provides a `$.fetch()` API similar to the Fetch API in browsers.

I feel okay with either of the following:

1. Having a disabled-by-default built-in rule that calls AmputatorBot - and I hope there can be some heuristics that allows skipping the call to AmputatorBot for most non-AMP pages.
2. Having a built-in rule that implements the proper rules locally, but maybe still disabled by default if it's not going to work in 100% of cases.
3. Keep the rule as a custom one and people can get it from you/somewhere else.

kj4ezj commented:
You are right, I remember seeing `$.fetch()` now. It had been a few weeks, I forgot about that.

Awesome, thank you for discussing it with me! I'll look at implementing number 1 or number 2 when I have some time and open another pull request. Cool project BTW, this is definitely a feature Android has needed for some time.

kj4ezj commented:
I opened [issue 25](https://github.com/zhanghai/Untracker/issues/25) to track this feature request.
